edb/schema/schema.py

```diff
@@ -1004,19 +1284,401 @@ def __init__(
+        return (
+            self._base_schema.get_referrers(
+                scls,
+                scls_type=scls_type,
+                field_name=field_name,
+            )
+            | self._top_schema.get_referrers(
+                scls,
+                scls_type=scls_type,
+                field_name=field_name,
+            )
+        )
+
+    def get_referrers_ex(
+        self,
+        scls: so.Object,
+    ) -> Dict[
+        Tuple[Type[so.Object], str],
+        FrozenSet[so.Object],
+    ]:
+        base = self._base_schema.get_referrers_ex(scls)
+        top = self._top_schema.get_referrers_ex(scls)
+        return {
+            k: base.get(k, frozenset()) | top.get(k, frozenset())
+            for k in itertools.chain(base, top)
+        }
+
+    @overload
+    def get_by_id(
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object, so.NoDefaultT] = so.NoDefault,
+        *,
+        type: None = None,
+    ) -> so.Object:
+        ...
+
+    @overload
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object_T, so.NoDefaultT] = so.NoDefault,
+        *,
+        type: Optional[Type[so.Object_T]] = None,
+    ) -> so.Object_T:
+        ...
+
+    @overload
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: None = None,
+        *,
+        type: None = None,
+    ) -> Optional[so.Object]:
+        ...
+
+    @overload
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: None = None,
+        *,
+        type: Type[so.Object_T],
+    ) -> Optional[so.Object_T]:
+        ...
+
+    def get_by_id(  # NoQA: F811
+        self,
+        obj_id: uuid.UUID,
+        default: Union[so.Object_T, so.NoDefaultT, None] = so.NoDefault,
+        *,
+        type: Optional[Type[so.Object_T]] = None,
+    ) -> Optional[so.Object_T]:
+        try:
+            return self._top_schema.get_by_id(
+                obj_id, type=type)
+        except errors.InvalidReferenceError:
+            # seems like a mypy bug below
+            return self._base_schema.get_by_id(  # type: ignore
+                obj_id, default=default, type=type)
+
+    @overload
+    def get_global(
+        self,
+        objtype: Type[so.Object_T],
+        name: str,
+        default: Union[so.Object_T, so.NoDefaultT] = so.NoDefault,
+    ) -> so.Object_T:
+        ...
+
+    @overload
+    def get_global(  # NoQA: F811
+        self,
+        objtype: Type[so.Object_T],
+        name: str,
+        default: None = None,
+    ) -> Optional[so.Object_T]:
+        ...
+
+    def get_global(  # NoQA: F811
+        self,
+        objtype: Type[so.Object_T],
+        name: str,
+        default: Union[so.Object_T, so.NoDefaultT, None] = so.NoDefault,
+    ) -> Optional[so.Object_T]:
+        try:
+            return self._top_schema.get_global(objtype, name)
+        except errors.InvalidReferenceError:
```

Comment:
Ditto, optimize this one too? Checking `_globalname_to_id` on `top_schema` should do it